### DISCUSSION TITLE
LDPC BP Decoder as RNN?

### DISCUSSION THREAD

**Discussion by Inv4lidn4m3** [NVlabs/sionna#8] [General]
Hello everyone,

First of all I would like to thank you for this initiative. This is a great idea!

With my coworkers we are studying the usage of neural networks and machine learning for digital signal processing, and more specifically in the context of linear block coding/decoding. We have a BP algorithm implemented as a RNN, see the corresponding implementation in [this paper](https://ieeexplore.ieee.org/document/9482479) (I am one of the author, Louis-Adrien Dufrène).

This implementation is fully differentiable and generic in the sens that you can described any factor graph in the RNN cell. Moreover, the BP algorithm is implemented with matrix manipulations, hence more complient with the "Tensorflow spirit". 

Some modifications of our code are needed to support all the features already proposed by the Sionna's LDPC Decoder, but since the main basis is there, it should be managable.

Would you be interested in this architecture?

Edit (01/10/23) : We have published a paper in IEEE Transactions on Communications (https://ieeexplore.ieee.org/document/9896912) on automatic code discovery using our decoder architecture. The code to replicate the paper results is available in open source here https://github.com/Orange-OpenSource/GNBP. The code do not use Sionna at the moment because of legacy development, and we are wondering how we could actually add some Sionna's feature within the notebook to apply for the "Made with Sionna" section. Could you tell us how much of Sionna's features would we need to add? Does a simple Sionna's AWGN layer be enough?

LA

**Comment by SebastianCa** [comment]
It is great to hear about your work.

We are currently thinking about various ways to promote Sionna-based research. Could you develop and share a Sionna-based Jupyter notebook that contains your RNN decoder and reproduces some results from your paper for this purpose?

**Reply by Inv4lidn4m3** [reply]
Hello Sebastian,

Thanks for the approval 😃 

On our side, we are currently working on several paper publications and we plan to publish the whole code on the Orange Organization's Github. Once our code is published and if you are still interested, we would be able to provide a notebook based on our code & Sionna. Consequently, we still need some time before being able to contribute to Sionna.

I will keep you in touch of our internal process.

Does this work for you?

**Reply by SebastianCa** [reply]
Hello Louis-Adrien,

This sounds great. We are looking forward to your publication.

**Reply by Inv4lidn4m3** [reply]
Hello Sebastian,

I talked internally about the possibility to contribute our code to Sionna. I almost have the go. Orange just ask if the code can be published under a MIT or BDS3 licence, instead of the Apache 2 one?

**Reply by SebastianCa** [reply]
Hi Louis-Adrien,
 
We plan to introduce a new section on the website called “Made with Sionna” with one of the next versions of Sionna.
 
The idea is to present external work that has been done with Sionna and directly link to example notebooks similar to the tutorial section.
One advantage is that you can publish the notebook in a separate repository (or any other online accessible folder) under your own license (e.g., MIT or BSD-3). We would then add a short summary of and link to your work in this section (including credits to the authors, paper reference etc.).

**Reply by Inv4lidn4m3** [reply]
This is nice! I think this is the best way to do this contribution. It will be easier for us to propose a separated notebook with Sionna + our work, providing visibility to both, while allowing us to choose the licence.

We keep in touch through this thread!
Thanks!
